[CK-Tile] Merge transpose examples

tenpercent opened the pull request:
## Proposed changes

`example/ck_tile/{35_batched_transpose,37_transpose}` implement the same transposition operator, but invoke different pipelines. We merge them into `35_batched_transpose` example and add dispatch on the pipeline

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

tenpercent committed:
unify pipeline signature with existing example

tenpercent committed:
iwyu

tenpercent committed:
move stuff around in load-tile-transpose

tenpercent committed:
cleanups in batched transpose pipeline

tenpercent committed:
comments

tenpercent committed:
use same inputs size

tenpercent committed:
cleaner printf

tenpercent committed:
print host args

tenpercent committed:
use 64 block sides in the 37_transpose example

tenpercent committed:
roll back grid dimension size adjustment for 37_transpose example

tenpercent committed:
transpose grid for 37_transpose to unify with 35_batched_transpose

tenpercent committed:
unify grid computation logic

tenpercent committed:
make policy methods device only (since they are used only on device from the pipeline)

tenpercent committed:
more host/device attribute cleanups

tenpercent committed:
copy over problem

tenpercent committed:
move over pipeline and policy

tenpercent committed:
add switch to batched transpose api

tenpercent committed:
make the lds problem more similar to original problem

tenpercent committed:
factor out logic into traits

tenpercent committed:
factor out conditional compilation into trait parameter

tenpercent committed:
propagate pipeline to args

tenpercent committed:
unhardcode pipeline dispatch parameter

ThomasNing commented on example/ck_tile/35_batched_transpose/batched_transpose_api.cpp:
```diff
@@ -53,21 +101,28 @@ float batched_transpose_dispatch(batched_transpose_kargs& a, ck_tile::stream_con
-    F(bf16, ck_tile::bf16_t, 64, 64, 64, 64, 8, 8, true, true)   \
-    F(bf16, ck_tile::bf16_t, 64, 64, 64, 64, 8, 8, false, false)
+#define FOREACH_TRANSPOSE_PARAM(F)                                  \
+    F(fp8, ck_tile::fp8_t, 64, 64, 64, 64, 8, 8, true, true, 0)     \
```
Instead of having the warp_x and warp_y, we could have how many warps initialized inside a block instead of having a warp tile size here.

tenpercent replied:
Moved them to the Problems' bodies

ThomasNing commented on example/ck_tile/37_transpose/batched_transpose_kernel.hpp:
```diff
@@ -114,7 +113,7 @@ struct BatchedTransposeKernel
             {static_cast<ck_tile::index_t>(iN), static_cast<ck_tile::index_t>(iM)});
 
-        Pipeline{}(x_block_window, y_block_window, smem);
+        Pipeline{}(x_block_window, y_block_window);
```
Could I know why we delete the smem pointer here?

tenpercent replied:
The `smem` pointer can be moved to the pipeline's `operator()` internals and doesn't have to be in the function interface, so I think it's cleaner this way

tenpercent replied:
Actually, I think I can roll back any changes to `37_transpose` since I copied the modified logic over to `ops`

ThomasNing replied:
Sure, if putting it inside the pipeline works, then we can put it in the pipeline.

ThomasNing commented on include/ck_tile/ops/batched_transpose/pipeline/batched_transpose_lds_policy.hpp:
```diff
@@ -0,0 +1,151 @@
+    template <typename Problem>
+    CK_TILE_DEVICE static constexpr auto GetVectorSize()
+    {
+        return 16 / sizeof(typename Problem::DataType);
```
This one should be a template depending on the datatype and if it is padding.

tenpercent replied:
It looks like it's already available as a part of each Problem - need to refactor

ThomasNing commented on include/ck_tile/ops/batched_transpose/pipeline/batched_transpose_lds_policy.hpp:
```diff
@@ -0,0 +1,151 @@
+    CK_TILE_DEVICE static constexpr auto MakeInputDistribution()
+    {
+        constexpr index_t BlockSize         = Problem::kBlockSize;
+        constexpr index_t LeadDimPerBlock   = Problem::kLeadSizePerBlock;
```
This one should be corresponding to the block tile information.

ThomasNing commented on include/ck_tile/ops/batched_transpose/pipeline/batched_transpose_lds_policy.hpp:
```diff
@@ -0,0 +1,151 @@
+    {
+        constexpr index_t kLeadDimPerBlock   = Problem::kLeadSizePerBlock;
+        constexpr index_t kSecondDimPerBlock = Problem::kSecondSizePerBlock;
+        constexpr index_t kVectorSize        = 16 / sizeof(typename Problem::DataType);
```
This corresponds to the kVectorLoadSize.

tenpercent replied:
Started using VectorLoadSize from the Problem internals

ThomasNing commented on include/ck_tile/ops/batched_transpose/pipeline/batched_transpose_lds_policy.hpp:
```diff
@@ -0,0 +1,151 @@
+        constexpr index_t kLeadDimPerBlock   = Problem::kLeadSizePerBlock;
+        constexpr index_t kSecondDimPerBlock = Problem::kSecondSizePerBlock;
+
+        constexpr index_t kVectorSize = 8 / sizeof(typename Problem::DataType);
```
Please align that with other kVectorSize.

tenpercent replied:
Started using VectorLoadSize from the Problem internals

ThomasNing commented on include/ck_tile/ops/batched_transpose/pipeline/batched_transpose_lds_problem.hpp:
```diff
@@ -0,0 +1,103 @@
+template <typename DataType_,
+          typename BlockTile, // sequence<block_x, block_y>
+          typename WarpTile>  // sequence<warp_x, warp_y>
+struct BatchedTransposeLdsProblem
```
The LDS transpose problem and thread level transpose problem could be merged.

tenpercent replied:
They could, but the problem parts are called different between the problem structs, and the LDS struct has a few extra parameters related to LDS tiles - so they are sufficiently different to be kept in separate structs, I think? Anyways, I can at least try to have the consistent names for the same entities between the structs

ThomasNing replied:
Yes, we could have a base part and variation part.

ThomasNing commented on include/ck_tile/ops/batched_transpose/pipeline/batched_transpose_lds_policy.hpp:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    template <typename Problem>
+    CK_TILE_DEVICE static constexpr auto MakeInputDistribution()
```
MakeInputDistribution could also be merged with the thread level transpose method.

tenpercent replied:
factored it out

tenpercent committed:
refactor vector size

tenpercent committed:
put warp tile out of dispatch

tenpercent committed:
rename template parameter for trait

tenpercent committed:
rewrite vector size in terms of problem

tenpercent committed:
mark policy-internal struct variable as device

tenpercent committed:
factor out input distribution and thread access pattern from policies

tenpercent committed:
reword vector size

tenpercent committed:
use datatype across batched transpose pipelines, problems and kernel

tenpercent committed:
remove transpose traits from lds pipeline

tenpercent committed:
add padding to the lds pipeline *interface*

tenpercent committed:
add comment

ThomasNing commented:
Could we delete the transpose folder operator? Thanks!

tenpercent committed:
remove ck_tile example #37

tenpercent committed:
update cmakelists

tenpercent commented:
Deleted ck_tile example 37. Padding doesn't work in the LDS pipeline, we can address it in the next PR?

tenpercent committed:
add test for new pipeline

tenpercent commented:
Added tests for the new pipeline and a few cases are failing (numerical mismatch), with failures related to incorrectly applied padding

tenpercent committed:
update batched transpose test

tenpercent committed:
roll back load_tile_transpose changes

tenpercent committed:
remove comments

tenpercent committed:
pack dispatch parameters into a config

tenpercent committed:
padM can be enabled

tenpercent committed:
adjust lds vector size to enable padding along N

tenpercent committed:
update test

tenpercent committed:
Merge branch 'develop' into tenpercent/transpose

tenpercent commented:
All tests pass now

tenpercent committed:
clean up logic

tenpercent committed:
Merge branch 'develop' into tenpercent/transpose

tenpercent committed:
swap m/n input vector size

tenpercent committed:
Merge branch 'tenpercent/transpose' of https://github.com/rocm/composable_kernel into tenpercent/transpose

tenpercent committed:
Merge remote-tracking branch 'origin/develop' into tenpercent/transpose

tenpercent committed:
adjust perf test script

tenpercent committed:
sweep over C/W in perf test

tenpercent committed:
count both read and written bytes into bandwidth (x2 the number)

tenpercent committed:
clang-format

tenpercent committed:
widen size range for perf test

tenpercent committed:
remove 64k x 64k case; it's too large for index

tenpercent committed:
remove thread tile from dispatch

ThomasNing committed:
Solve merge conflict

tenpercent committed:
fix compile

ThomasNing committed:
modify the transpose

ThomasNing committed:
solve the test error and clang format

tenpercent committed:
Merge remote-tracking branch 'origin/develop' into tenpercent/transpose

aosewski committed:
Add v3 support for Groupd fwd conv+bias+clamp & ckProfiler (#2463)

* Add logging to IsSupported.

* Less casting in AddClamp

* Conv+bias+clamp instances & profiler BF16

* Fix 3D instances & run just 1x for verification.

* :Run just once for verification conv fwd.

* ckProfiler conv fwd clampwq

* Remove exec bit & formatting

* Add support for MultiD for grouped conv fwd v3.

* Enable 2Lds.

* clean

* align instances

* align instances

* profiler fixes

* Fixes

* fix

* fix

---------

Co-authored-by: Adam Osewski <root@quanta-ccs-aus-f01-19.cs-aus.dcgpu>
Co-authored-by: Bartłomiej Kocot <barkocot@amd.com>

rahjain-amd committed:
Fixing 0ms and inf GB/s issue in img2col (#2565)

issue :
====
``` sh
$ bin/tile_example_img2col
Perf: 0 ms, inf GB/s
```

solution :
======
Problem occured because config.time_kernel is false by default.
if false, then no need to calculate perf, just print proper message

`image_to_coloumn: pass, No Perf generated due to config.time_kernel=0`

ThomasNing committed:
Merge branch 'develop' into tenpercent/transpose

ThomasNing committed:
merge with develop

ThomasNing committed:
solve clang format
